### PR TITLE
Agnostic ColumnSelector

### LINKED ISSUE #658: [FEATURE] Narwhals migration for dataframe-agnostic codebase
# Description

Creating this issue to keep track of which classes/function could benefit from adopting Narwhals.

| Class/Function/Module | Status | Related | Notes |
| - | - | - | - |
|  `preprocessing.ColumnDropper` | ✅ | Solved in #655 | |
|  `preprocessing.ColumnSelector` | ✅  | Solved in #659 | | 
|  `preprocessing.PandasTypeSelector` | ✅ | Solved in #670  |  Consider creating another class `TypeSelector` and deprecate this one |
|  `common.TrainOnlyTransformerMixin` |🔲 |  | Uses specific pandas hashing functionality. I wonder how crucial is to hash the index as well. If it's not we could just use `.to_numpy()` and hash the array data? |
|  `model_selection.TimeGapSplit` | ✅ | Solved in #668  |  |
|  `model_selection.GroupedTimeSeriesSplit` | 🔲|  | Related to #605  |
|  `projections.InformationFilter` | ✅ | Solved in [commit](https://github.com/koaning/scikit-lego/commit/0773db9325e87041519aed79371ee30955e0028f) | |
|  `meta.RegressionOutlierDetector` | ✅| Solved in #665  | |
|  `meta.hierarchical_predictor.py` | ✅| Solved in #667  |  |
|  `meta.grouped_transformer.py` | ✅| Solved in #667  |  |
|  `meta.grouped_predictor.py` | ✅| Solved in #667  | |
|  `linear_models._FairClassifier` | ✅ | Solved in #669  |  |
|  `pandas_utils.py` | 🚧| Partially #661  | (*) |
|  `datasets.py` | 🚫|  | It would require `read_csv` function |

Personally I would wait to have at least `preprocessing.pandastransformers.py` entire migration before bumping to v0.9.0.

cc @MarcoGorelli @anopsy 

## (*) Regarding `pandas_utils`

Changing `log_step` to narwhals is fairly easy ([~4 lines of code](https://github.com/koaning/scikit-lego/blob/feat/narwhals-utils/sklego/pandas_utils.py#L95)), however as this decorator is supposed to work for _any_ function that operates on pandas, doing so would limit its functionality. It could be reasonable to add another one which although restricted to narwhals methods, it can interoperate with all its compatible dataframes.

**Legend**

✅ Done 
🚧 WIP 
🔲 Not Started 
🚫 Won't do

### PR DESCRIPTION
**Pull request by anopsy**
# Description

Narwhalyfied ColumnSelector, I added an example with polars df to the docstrings and adjusted the test_columnselector.py
I tried to test the docstrings, but it doesn't work. Maybe we could add those to sklego in the future?

Fixes #658 - partly

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist:

- [x] My code follows the style guidelines (ruff)
- [ ] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation (also to the readme.md)
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added tests to check whether the new feature adheres to the sklearn convention
- [x] New and existing unit tests pass locally with my changes

If you feel your PR is ready for a review, ping @FBruzzesi or @koaning.

### CODE REVIEW

File: sklego/preprocessing/pandastransformers.py [hunk 1]
```diff
@@ -329,52 +344,68 @@ class ColumnSelector(BaseEstimator, TransformerMixin):
...
+    import polars as pl
+    from sklego.preprocessing import ColumnSelector
+
+    df_pl = pl.DataFrame({
+        "name": ["Swen", "Victor", "Alex"],
+        "length": [1.82, 1.85, 1.80],
+        "shoesize": [42, 44, 45]
+    })
+
+    ColumnSelector(["length", "shoesize"]).fit_transform(df_pl)
```
Comment (MarcoGorelli) [inline sklego/preprocessing/pandastransformers.py:371]: what you show below here doesn't look like Polars output, I think that may need updating?

File: sklego/preprocessing/pandastransformers.py [hunk 2]
```diff
@@ -404,40 +435,41 @@ def fit(self, X, y=None):
...
         ValueError
-            If dropping the specified columns would result in an empty output DataFrame.
+            If provided list of columns to select is empty and would result in an empty output DataFrame.
         """
         self.columns_ = as_list(self.columns)
-        self._check_X_for_type(X)
-        self._check_column_length()
+        X = nw.from_native(X)
         self._check_column_names(X)
+        self.feature_names_ = [x for x in X.columns if x in self.columns_]
```
Comment (MarcoGorelli) [inline sklego/preprocessing/pandastransformers.py:447]: Why is this line needed?

File: sklego/preprocessing/pandastransformers.py [hunk 3]
```diff
@@ -404,40 +435,41 @@ def fit(self, X, y=None):
...
         TypeError
-            If `X` is not a `pd.DataFrame` object.
+            If `X` is not a supported DataFrame.
         """
-        self._check_X_for_type(X)
+        X = nw.from_native(X)
         if self.columns:
-            return X[self.columns_]
-        return X
+            return nw.to_native(X.select(nw.col(self.columns_)))
```
Comment (MarcoGorelli) [inline sklego/preprocessing/pandastransformers.py:471]: bit simpler, you can just do `X.select(self._columns_)`

**Review by MarcoGorelli — COMMENTED** [review]
that was fast! just left some initial comments ahead of maintainers' reviews

File: sklego/preprocessing/pandastransformers.py (hunk 2 above)
Comment (MarcoGorelli) [reply sklego/preprocessing/pandastransformers.py:447]: I think you don't need `self.feature_names_` at all, it's not in the current code

File: sklego/preprocessing/pandastransformers.py (hunk 1 above)
Comment (anopsy) [reply sklego/preprocessing/pandastransformers.py:371]: Let me fix that

Comment (anopsy) [reply sklego/preprocessing/pandastransformers.py:371]: image
So I used polars output but it renders this way in mkdocs:

image

Comment (MarcoGorelli) [reply sklego/preprocessing/pandastransformers.py:371]: ah I think it's just a font that needs setting, the default one isn't exactly monospace

File: sklego/preprocessing/pandastransformers.py (hunk 2 above)
Comment (anopsy) [reply sklego/preprocessing/pandastransformers.py:447]: indeed it's not in Attributes list, I see now that the fit performs only checks, cool thank you

File: sklego/preprocessing/pandastransformers.py (hunk 1 above)
Comment (anopsy) [reply sklego/preprocessing/pandastransformers.py:371]: Is that something I can fix or not really @FBruzzesi ?

Comment (FBruzzesi) [reply sklego/preprocessing/pandastransformers.py:373]: I never rendered that! @MarcoGorelli send help 😂!

Comment (MarcoGorelli) [reply sklego/preprocessing/pandastransformers.py:373]: 😄 dunno, in Narwhals I just have
```
theme:
  name: material
  font: false
```

Comment (koaning) [reply sklego/preprocessing/pandastransformers.py:373]: I recall doing stuff like that in the past as well, it's a monofont isnt always mono kind of a thing. Especially when you add emojis or fancy kinds of unicode to the mix.

Comment (anopsy) [reply sklego/preprocessing/pandastransformers.py:373]: image
looks great now! I'll push it then

**Review by FBruzzesi — APPROVED** [review]
Thanks a lot! I start to like this a lot!

NiceGIF


P.S. Test do not run because it is not a PR to `main` branch, but I can confirm that it is all good (locally):
image

**Comment by anopsy** [comment]
Dude I've run those tests like 13 times, and also full tests twice (btw you can make lunch while they're running, it takes so long) . Thank you team for the support! I'm so excited

**Comment by anopsy** [comment]
Could we also want some doctests? Or that's too boujee 😅
